data/random-battles/gen9/teams.ts

```diff
@@ -2356,6 +2360,303 @@ export class RandomTeams {
 		return team;
 	}
 
+	randomFactorySets: {[format: string]: {[species: string]: BattleFactorySpecies}} = require('./factory-sets.json');
+
+	randomFactorySet(
+		species: Species, teamData: RandomTeamsTypes.FactoryTeamDetails, tier: string
+	): RandomTeamsTypes.RandomFactorySet | null {
+		const id = toID(species.name);
+		const setList = this.randomFactorySets[tier][id].sets;
+
+		const itemsLimited = ['choicespecs', 'choiceband', 'choicescarf'];
+		const movesLimited: {[k: string]: string} = {
+			stealthrock: 'stealthRock',
+			stoneaxe: 'stealthRock',
+			spikes: 'spikes',
+			ceaselessedge: 'spikes',
+			toxicspikes: 'toxicSpikes',
+			rapidspin: 'hazardClear',
+			defog: 'hazardClear',
+		};
+		const abilitiesLimited: {[k: string]: string} = {
+			toxicdebris: 'toxicSpikes',
+		};
+
+		// Build a pool of eligible sets, given the team partners
+		// Also keep track of moves and items limited to one per team
+		const effectivePool: {
+			set: BattleFactorySet, moves?: string[], item?: string,
+		}[] = [];
+
+		for (const curSet of setList) {
+			let reject = false;
+
+			// limit to 1 dedicated tera user per team
+			if (curSet.wantsTera && teamData.wantsTeraCount) {
+				continue;
+			}
+
+			// reject disallowed items, specifically a second of any given choice item
+			const allowedItems: string[] = [];
+			for (const itemString of curSet.item) {
+				const item = this.dex.items.get(itemString);
+				if (itemsLimited.includes(item.id) && teamData.has[item.id]) continue;
+				allowedItems.push(itemString);
+			}
+			if (allowedItems.length === 0) continue;
+			const curSetItem = this.sample(allowedItems);
+
+			const curSetAbility = this.sample(curSet.ability);
+			const ability = this.dex.abilities.get(curSetAbility);
+
+			if (abilitiesLimited[ability.id] && teamData.has[abilitiesLimited[ability.id]]) continue;
+
+			const curSetMoves: string[] = [];
+			for (const move of curSet.moves) {
+				const allowedMoves: string[] = [];
+				for (const m of move) {
+					const moveId = toID(m);
```

Comment:
Why not also use `toID` for items and abilities? Or, since you control the set data, you can just use the names directly, like this:
```typescript
		const abilitiesLimited: {[k: string]: string} = {
			'Toxic Debris': 'toxicSpikes',
		};
```

Reply:
Done, thanks!